### ISSUE TITLE
Add mechanism for tool dependencies.

### ISSUE THREAD

**Issue by victordibia** [microsoft/autogen#4263]

**Comment by ekzhu** [comment]
This can be helped by using distributed agent runtime. Agents don't have to be in the same application process. The agent that uses tools can be running from a separate machine. 

We can make agentchat to run on distributed runtime.

**Comment by victordibia** [comment]
I agree that the challenge above can be helped by using a distributed runtime.  
We will need to create a design doc/discussion around how to support distributed runtimes in AgentChat. 

Beyond that, any other comments on the proposed implementation (I added a _how_ section.) ?

**Comment by ekzhu** [comment]
I am wondering if the dependencies & secret should be provided and resolved at start of a runtime in the backend rather than managed through some UI. It is very hard to secure the application environment against arbitrary code execution (i.e., user defined tools).

**Comment by victordibia** [comment]
I agree that end users should not be defining tools.
However - AGS is a _developer tool_. For debugging. Fair to assume similar exposure with e.g., a notebook?
My feeling is that it should be treated sort of like an IDE? 
Open to thoughts here.
